tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,392 @@
+                    // TRUE with it being previously false and followed by that we are
+                    // initializing the unsafe Cell field with the value
+                    unsafe {
+                        self.set_value(value);
+                    }
+
+                    Ok(())
+                } else {
+                    Err(SetError::AlreadyInitializedError(value))
+                }
+            }
+            Err(_) => {
+                // If we failed to lock the mutex, it means some other task is
+                // trying to set the value, so we return an error.
+                Err(SetError::InitializingError(value))
+            }
+        }
+    }
+
+    /// Takes the value from the cell, destroying the cell in the process.
+    /// Returns `None` if the cell is empty.
+    pub fn into_inner(mut self) -> Option<T> {
+        if self.initialized() {
+            // Since we have taken ownership of self, its drop implementation
+            // will be called by the end of this function, to prevent a double
+            // free we will set the value_set to false so that the drop
+            // implementation does not try to drop the value again.
+            *self.value_set.get_mut() = false;
+
+            // SAFETY: The SetOnce is currently initialized, we can assume the
+            // value is initialized and return that, when we return the value
+            // we give the drop handler to the return scope.
+            Some(unsafe { self.value.with_mut(|ptr| ptr::read(ptr).assume_init()) })
+        } else {
+            None
+        }
+    }
+
+    /// Waits until set is called. The future returned will keep blocking until
+    /// the `SetOnce` is initialized.
+    ///
+    /// If the `SetOnce` is already initialized, it will return the value
+    // immediately.
+    ///
+    /// # Panics
+    ///
+    /// If the `SetOnce` is not initialized after waiting, it will panic. To
+    /// avoid this, use `get_wait()` which returns an `Option<&T>` instead of
+    /// `&T`.
+    pub async fn wait(&self) -> &T {
+        match self.get_wait().await {
+            Some(val) => val,
+            _ => panic!("SetOnce::wait called but the SetOnce is not initialized"),
+        }
+    }
+
+    /// Waits until set is called.
+    ///
+    /// If the state failed to initialize it will return `None`.
+    pub async fn get_wait(&self) -> Option<&T> {
+        let notify_fut = self.notify.notified();
+        pin!(notify_fut);
+
+        if self.value_set.load(Ordering::Acquire) {
+            // SAFETY: the state is initialized
+            return Some(unsafe { self.get_unchecked() });
+        }
+        // wait until the value is set
+        (&mut notify_fut).await;
+
+        // look at the state again
+        if self.value_set.load(Ordering::Acquire) {
+            // SAFETY: the state is initialized
+            return Some(unsafe { self.get_unchecked() });
+        }
+
+        None
```

Comment:
No need to check this. This also lets you get rid of `get_wait` and just have `wait`. It's not possible for the state to fail to initialize.
```suggestion
        Some(unsafe { self.get_unchecked() })
```